File: hackit/wireless/plugins/ruby/scripts/airbase_ng.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'
require 'open3'
require 'tempfile'

def emit(event, iface, bssid, ssid, channel, data)
  puts JSON.generate(event: event, iface: iface, bssid: bssid, ssid: ssid, channel: channel, data: data, timestamp: Time.now.iso8601)
  $stdout.flush
end

def write_hostapd_conf(path, iface, ssid, passphrase, channel)
  File.write(path, <<~CONF)
    interface=#{iface}
    driver=nl80211
    ssid=#{ssid}
    hw_mode=g
    channel=#{channel}
    wpa=2
    wpa_passphrase=#{passphrase}
    wpa_key_mgmt=WPA-PSK
    wpa_pairwise=TKIP CCMP
    rsn_pairwise=CCMP
    auth_algs=1
    macaddr_acl=0
    logger_stdout=-1
    logger_stdout_level=2
  CONF
end

def write_dnsmasq_conf(path, iface)
  subnet = "10.0.0"
  File.write(path, <<~CONF)
    interface=#{iface}
    dhcp-range=#{subnet}.10,#{subnet}.100,255.255.255.0,12h
    dhcp-option=3,#{subnet}.1
    dhcp-option=6,8.8.8.8,1.1.1.1
    server=8.8.8.8
    server=1.1.1.1
    no-resolv
    log-queries
    log-dhcp
    bind-dynamic
  CONF
end

begin
  iface = ARGV[0]
  ssid = ARGV[1] || 'RogueAP'
  channel = (ARGV[2] || '6').to_i
  passphrase = ARGV[3] || 'roguepass123'

  raise 'interface required' unless iface

  hostapd_conf = Tempfile.new(['hostapd', '.conf'])
  dnsmasq_conf = Tempfile.new(['dnsmasq', '.conf'])

  write_hostapd_conf(hostapd_conf.path, iface, ssid, passphrase, channel)
  write_dnsmasq_conf(dnsmasq_conf.path, iface)

  emit('airbase_start', iface, '', ssid, channel, { passphrase: passphrase })

  airbase = `which airbase-ng 2>/dev/null`.strip
  hostapd_bin = `which hostapd 2>/dev/null`.strip
  dnsmasq_bin = `which dnsmasq 2>/dev/null`.strip

  threads = []

  if airbase.empty? && hostapd_bin.empty?
    raise 'neither airbase-ng nor hostapd found in PATH'
  end

  if !airbase.empty?
    threads << Thread.new do
      cmd = [airbase, '-a', '-e', ssid, '-c', channel.to_s, '-W', '1', iface]
      emit('airbase_launch', iface, '', ssid, channel, { command: cmd.join(' ') })
      Open3.popen3(*cmd) do |stdin, stdout, stderr, thread|
        stdin.close
        [stdout, stderr].each do |r|
          Thread.new do
            r.each_line do |line|
              emit('airbase_output', iface, '', ssid, channel, { line: line.strip })
            end
          end
        end
        thread.join
      end
    end
  end

  if !hostapd_bin.empty?
    threads << Thread.new do
      cmd = [hostapd_bin, hostapd_conf.path]
      emit('hostapd_launch', iface, '', ssid, channel, { command: cmd.join(' ') })
      Open3.popen3(*cmd) do |stdin, stdout, stderr, thread|
        stdin.close
        [stdout, stderr].each do |r|
          Thread.new do
            r.each_line do |line|
              emit('hostapd_output', iface, '', ssid, channel, { line: line.strip })
            end
          end
        end
        thread.join
      end
    end
  end

  if !dnsmasq_bin.empty?
    threads << Thread.new do
      cmd = [dnsmasq_bin, '-C', dnsmasq_conf.path, '--no-daemon']
      emit('dnsmasq_launch', iface, '', ssid, channel, { command: cmd.join(' ') })
      Open3.popen3(*cmd) do |stdin, stdout, stderr, thread|
        stdin.close
        [stdout, stderr].each do |r|
          Thread.new do
            r.each_line do |line|
              emit('dnsmasq_output', iface, '', ssid, channel, { line: line.strip })
            end
          end
        end
        thread.join
      end
    end
  end

  threads.each(&:join)

rescue Interrupt
  emit('airbase_interrupted', iface || '', '', ssid || '', channel || 0, {})
rescue => e
  $stderr.puts JSON.generate(event: 'error', data: { message: e.message, backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601)
  exit 1
ensure
  hostapd_conf&.unlink
  dnsmasq_conf&.unlink
end

File: hackit/wireless/plugins/lua/scripts/airbase_ng.lua
local iface = arg[1]
local ssid = arg[2] or ("AP_" .. tostring(math.floor(os.time() * 1000 % 10000)))
local channel = tonumber(arg[3]) or 6
local wpa2_pass = arg[4] or "password123"

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":' .. json_escape(iface) .. ',"bssid":"","ssid":' .. json_escape(ssid) .. ',"channel":' .. tostring(channel) .. ',"data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local function write_configs()
  local hostapd_conf = "/tmp/hostapd_" .. iface .. ".conf"
  local dnsmasq_conf = "/tmp/dnsmasq_" .. iface .. ".conf"
  local f = io.open(hostapd_conf, "w")
  if not f then
    telemetry("error", '{"error":"Cannot write hostapd config"}')
    return nil, nil
  end
  f:write("interface=" .. iface .. "\n")
  f:write("driver=nl80211\n")
  f:write("ssid=" .. ssid .. "\n")
  f:write("hw_mode=g\n")
  f:write("channel=" .. channel .. "\n")
  f:write("wpa=2\n")
  f:write("wpa_passphrase=" .. wpa2_pass .. "\n")
  f:write("wpa_key_mgmt=WPA-PSK\n")
  f:write("wpa_pairwise=TKIP CCMP\n")
  f:write("rsn_pairwise=CCMP\n")
  f:write("auth_algs=1\n")
  f:close()
  local d = io.open(dnsmasq_conf, "w")
  if not d then
    telemetry("error", '{"error":"Cannot write dnsmasq config"}')
    return hostapd_conf, nil
  end
  d:write("interface=" .. iface .. "\n")
  d:write("dhcp-range=10.0.0.10,10.0.0.100,12h\n")
  d:write("dhcp-option=3,10.0.0.1\n")
  d:write("dhcp-option=6,10.0.0.1\n")
  d:write("server=8.8.8.8\n")
  d:write("log-queries\n")
  d:write("log-dhcp\n")
  d:close()
  telemetry("config_written", '{"hostapd":' .. json_escape(hostapd_conf) .. ',"dnsmasq":' .. json_escape(dnsmasq_conf) .. '}')
  return hostapd_conf, dnsmasq_conf
end

local function setup_interface()
  os.execute("ip link set " .. iface .. " down 2>/dev/null")
  os.execute("ip addr add 10.0.0.1/24 dev " .. iface .. " 2>/dev/null")
  os.execute("ip link set " .. iface .. " up 2>/dev/null")
  telemetry("interface_setup", '{"ip":"10.0.0.1/24"}')
end

local function run_airbase()
  local cmd = "airbase-ng -e " .. ssid .. " -c " .. channel .. " -W 1 " .. iface .. " 2>/dev/null &"
  telemetry("airbase_start", '{"cmd":' .. json_escape(cmd) .. '}')
  os.execute(cmd)
  os.execute("sleep 1")
end

local function run_services(hostapd_conf, dnsmasq_conf)
  local cmd_h = "hostapd " .. hostapd_conf .. " 2>/dev/null &"
  local cmd_d = "dnsmasq -C " .. dnsmasq_conf .. " 2>/dev/null &"
  os.execute(cmd_h)
  os.execute(cmd_d)
  telemetry("services_started", '{"hostapd":"started","dnsmasq":"started"}')
end

local function capture_handshakes()
  local f = io.popen("tcpdump -i " .. iface .. " -nn -c 100 port 88 or port 67 or port 68 2>/dev/null", "r")
  if not f then return end
  local captured = 0
  while true do
    local line = f:read("*l")
    if not line then break end
    if line:find("EAPOL") or line:find("eapol") or line:find("SAP") then
      captured = captured + 1
      telemetry("handshake", '{"count":' .. tostring(captured) .. ',"detail":' .. json_escape(line:sub(1, 150)) .. '}')
    end
    if line:find("DHCP") then
      telemetry("dhcp_lease", '{"detail":' .. json_escape(line:sub(1, 150)) .. '}')
    end
  end
  f:close()
end

if not iface then
  telemetry("error", '{"error":"Usage: airbase_ng.lua <interface> [ssid] [channel] [wpa2_pass]"}')
  os.exit(1)
end

local ok, err = pcall(function()
  local hp_conf, dns_conf = write_configs()
  if not hp_conf then os.exit(1) end
  setup_interface()
  run_airbase()
  if dns_conf then run_services(hp_conf, dns_conf) end
  capture_handshakes()
end)

if not ok then
  telemetry("error", '{"error":' .. json_escape(tostring(err)) .. '}')
  os.exit(1)
end

telemetry("complete", '{"ssid":' .. json_escape(ssid) .. ',"channel":' .. tostring(channel) .. '}')
